### PR TITLE
Transition to pyproject.toml and setup.cfg

### LINKED ISSUE #1286: Transition Woodwork to use pyproject.toml and setup.cfg (moving away from setup.py)
> The future of Python packaging is pyproject.toml, and (for now) setup.cfg, based on [PEP 518](https://snarky.ca/clarifying-pep-518/) and (soon) [PEP 621](https://www.python.org/dev/peps/pep-0621/).

- http://ivory.idyll.org/blog/2021-transition-to-pyproject.toml-example.html
- Therefore we should migrate our Woodwork package to follow this new standard

## Step 1
- Create `pyproject.toml`:
```
[build-system]
requires = ["setuptools", "wheel"]
build-backend = 'setuptools.build_meta'
```

## Step 2
- Modify setup.cfg
```
[metadata]
name = woodwork
version = attr: woodwork.__version__
url = https://github.com/alteryx/woodwork
download_url = https://pypi.org/project/woodwork/
project_urls =
    Documentation = https://woodwork.alteryx.com
    Source = https://github.com/alteryx/woodwork
    Tracker = https://github.com/alteryx/woodwork/issues
author = Alteryx, Inc.
author_email = open_source_support@alteryx.com
classifier =
    Development Status :: 3 - Alpha
    Intended Audience :: Developers
    Intended Audience :: Science/Research'
    Programming Language :: Python :: 3
    Programming Language :: Python :: 3.8
    Programming Language :: Python :: 3.9
    Topic :: Software Development',
    Topic :: Scientific/Engineering',
    Operating System :: Microsoft :: Windows',
    Operating System :: POSIX',
    Operating System :: Unix',
    Operating System :: MacOS',
license = BSD 3-clause
description = a namespace for dataframes to help with typing in machine learning
long_description= file: README.md
keywords = data, science, machine learning, typing
platforms = any

[options]
zip_safe = False
install_requires =
    pandas >= 1.3.0
    scikit-learn >= 0.22
python_requires = >=3.8
include_package_data = True
packages = find:
package_dir=
    =woodwork


[options.extras_require]
update_checker =
    alteryx-open-src-update-checker >= 2.0.0
dask =
    dask[dataframe] >= 2021.10.0
koalas =
    pyspark >= 3.0.0, <3.2.0
    koalas >= 1.8.0, <=1.8.2
    pandas < 1.4.0
test =
    pip >= 21.3.1
    pytest == 6.0.1
    pytest-cov == 2.10.1
    pytest-xdist == 2.1.0
    mock == 4.0.3
    boto3 >= 1.10.45
    moto[all] >= 1.3.14
    smart-open >= 5.0.0
    pyarrow >= 4.0.1
complete =
    update_checker
    dask
    koalas

[options.package_data]
* =
    *.txt
    LICENSE
    README.md

[options.packages.find]
where=woodwork

[options.exclude_package_data]
* =
    * __pycache__
    *.py[co]

[options.entry_points]
console_scripts =
    woodwork = woodwork.__main__:cli

[aliases]
test=pytest

[flake8]
max-line-length = 88
extend-ignore = E203
exclude = docs/*
ignore = E501,W504,W503
per-file-ignores =
    **/__init__.py:F401
    **/tests/*:D

[tool:black]
target-version = ['py38', 'py39']

[tool:pytest]
addopts = --doctest-modules
python_files = woodwork/tests/*
filterwarnings =
    ignore::DeprecationWarning
    ignore::PendingDeprecationWarning

[isort]
profile=black
forced_separate=woodwork
multi_line_output=3
known_first_party=woodwork
skip=__init__.py
```

## Step 3
- Modify `setup.py`:
```
#!/usr/bin/env python
import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

## Step 4
- Remove all requirements files (not sure about dev-requirements)
- `dask-requirements.txt`
- `dev-requirements.txt`
- `koalas-requirements.txt`
- `requirements.txt`
- `test-requirements.txt`

## Step 5
- Verify wheel files are build properly, with correct versioning, and installs on all systems
- Modify `install.test.yml` so it has the following (notice the matrix.os
```
name: Install Test
jobs:
  install_ww_complete:
    name: ${{ matrix.python_version }} install woodwork complete
    strategy:
      fail-fast: false
      matrix:
        python_version: ["3.8", "3.9"]
        os: [ubuntu-latest, macos-latest, windows-latest]
    runs-on: ${{ matrix.os }}
```

## Step 6
- Determine how to handle latest dependency checker, and min dependency checker
- In addition, min-unit tests and latest unit tests (which should be fine?)

# Open Questions
## You can specify requirements.txt in setup.cfg with this
- https://pypi.org/project/setuptools-declarative-requirements/

## Do we need to use pip-tools
- https://github.com/jazzband/pip-tools

### Helpful Links
- https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html
- http://ivory.idyll.org/blog/2021-transition-to-pyproject.toml-example.html
- https://servilla.io/python-packaging-for-idiots-me/
- https://www.scivision.dev/minimal-setup-py-with-prerequisites/
- https://snarky.ca/what-the-heck-is-pyproject-toml/

### PR DESCRIPTION
**Pull request by jeff-hernandez**
Closes #1286

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,24 +1,132 @@
...
+project_urls =
+    Documentation = https://woodwork.alteryx.com
+    Source = https://github.com/alteryx/woodwork
+
+[options]
+packages = find:
+python_requires = >=3.7,<4
+include_package_data = True
+
+setup_requires =
```
Comment (gsheni) [inline setup.cfg:31]: This section is obsolete and it is not recommended that you have a `setup_requires`.
https://setuptools.pypa.io/en/latest/userguide/dependency_management.html

File: Makefile [hunk 2]
```diff
@@ -31,15 +31,15 @@ testcoverage:
 installdeps:
 	pip install --upgrade pip
 	pip install -e .
-	pip install -r dev-requirements.txt
+	pip install -e .[dev]
```
Comment (thehomebrewnerd) [inline Makefile]: You should wrap these install commands in quotes so they will also work for people using zsh: `pip install -e ".[dev]"`.

File: docs/source/release_notes.rst [hunk 3]
```diff
@@ -9,11 +9,12 @@ Release Notes
     * Fixes
         * Updated ``__str__`` output for ``Ordinal`` logical types (:pr:`1340`)
     * Changes
+        * Transition to use pyproject.toml and setup.cfg (moving away from setup.py) (:pr:`1346`)
     * Documentation Changes
     * Testing Changes
 
 .. Thanks to the following people for contributing to this release:
-    :user:`bchen1116`
+    :user:`bchen1116`, :user:`jeff-hernandez`
```
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:18]: Uncomment future release section

File: setup.cfg [hunk 4]
```diff
@@ -1,24 +1,128 @@
...
+    LICENSE
+    README.md
+
+[options.exclude_package_data]
+* =
+    * __pycache__
+    *.py[co]
+
+[bdist_wheel]
+universal = 1
```
Comment (thehomebrewnerd) [inline setup.cfg]: Remove this to prevent it from building a Python 2 wheel.

File: .github/workflows/build_docs.yml [hunk 5]
```diff
@@ -29,7 +29,7 @@ jobs:
           pip config --site set global.progress_bar off
           python -m pip install --upgrade pip
           python -m pip install -e .
-          python -m pip install -r dev-requirements.txt
+          python -m pip install -e .[dev]
```
Comment (thehomebrewnerd) [inline .github/workflows/build_docs.yml:32]: I think you can just do `python -m pip install .[dev]` here without the `python-m pip install -e .` command. Also, I don't think there is a need to install this in editable mode.

File: .github/workflows/latest_dependency_checker.yml [hunk 6]
```diff
@@ -25,7 +25,7 @@ jobs:
         source venv_core/bin/activate
         python -m pip install --upgrade pip
         python -m pip install -e .
```
Comment (thehomebrewnerd) [inline .github/workflows/latest_dependency_checker.yml:27]: I think this line might be unnecessary now.

File: .github/workflows/latest_dependency_checker.yml [hunk 7]
```diff
@@ -34,8 +34,8 @@ jobs:
         source venv_spark/bin/activate
         python -m pip install --upgrade pip
         python -m pip install -e .
```
Comment (thehomebrewnerd) [inline .github/workflows/latest_dependency_checker.yml:36]: Unnecessary?

File: .github/workflows/latest_dependency_checker.yml [hunk 8]
```diff
@@ -44,8 +44,8 @@ jobs:
         source venv_dask/bin/activate
         python -m pip install --upgrade pip
         python -m pip install -e .
```
Comment (thehomebrewnerd) [inline .github/workflows/latest_dependency_checker.yml:46]: Unnecessary?

File: .github/workflows/lint_check.yml [hunk 9]
```diff
@@ -29,7 +29,7 @@ jobs:
           python -m pip install --upgrade pip
           python -m pip install -e .[dask]
           python -m pip install -e .[spark]
-          python -m pip install -r dev-requirements.txt
-          python -m pip install -r test-requirements.txt
+          python -m pip install -e .[dev]
+          python -m pip install -e .[test]
```
Comment (thehomebrewnerd) [inline .github/workflows/lint_check.yml:33]: The dev requirements include everything, so I don't think you need any of the other install commands.

File: .github/workflows/unit_tests_with_latest_deps.yml [hunk 10]
```diff
@@ -37,7 +37,7 @@ jobs:
           python -m pip install -e unpacked_sdist/
       - name: Install woodwork - tests requirements
         run: |
-          python -m pip install -r unpacked_sdist/test-requirements.txt
+          python -m pip install -e unpacked_sdist/[test]
```
Comment (thehomebrewnerd) [inline .github/workflows/unit_tests_with_latest_deps.yml:40]: I don't think we need to install this in editable mode

File: .github/workflows/latest_dependency_checker.yml (hunk 6 above)
Comment (jeff-hernandez) [reply .github/workflows/latest_dependency_checker.yml:27]: are you referring to only line `python -m pip install -e .`

Comment (thehomebrewnerd) [reply .github/workflows/latest_dependency_checker.yml:27]: Yes, I think the `python -m pip instal .[test]` command should also install the library whereas before the extra `python -m pip install -r test-requirements.txt` command was needed to add the test requirements to the base library install.

Comment (jeff-hernandez) [reply .github/workflows/latest_dependency_checker.yml]: sounds good, updated

File: .github/workflows/build_docs.yml (hunk 5 above)
Comment (jeff-hernandez) [reply .github/workflows/build_docs.yml]: updated

File: .github/workflows/lint_check.yml (hunk 9 above)
Comment (jeff-hernandez) [reply .github/workflows/lint_check.yml]: updated

File: setup.cfg [hunk 11]
```diff
@@ -1,24 +1,125 @@
+[metadata]
+name = woodwork
+version = 0.14.0
+author = Alteryx Inc.
+license = BSD 3-clause
+long_description = file: READEME.md
+url = https://github.com/alteryx/woodwork/
+long_description_content_type = 'text/markdown'
```
Comment (thehomebrewnerd) [inline setup.cfg:8]: I don't know for sure if it matters, but we had trouble with this in Featuretools and ended up specifying this without any quotes: https://github.com/alteryx/featuretools/blob/aaa84e3634988b2e6d2c528ff240a397655024d7/setup.cfg#L31

File: setup.cfg [hunk 12]
```diff
@@ -1,24 +1,125 @@
...
+description = a two-dimensional data object with labeled axes and typing information
+
+classifiers =
+    Development Status :: 3 - Alpha
+    Intended Audience :: Developers
+    Programming Language :: Python :: 3
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
+    Programming Language :: Python :: 3.9
+    Programming Language :: Python :: 3.10
```
Comment (thehomebrewnerd) [inline setup.cfg:20]: Specify operating systems like in Featuretools?  https://github.com/alteryx/featuretools/blob/aaa84e3634988b2e6d2c528ff240a397655024d7/setup.cfg#L24

**Review by thehomebrewnerd — COMMENTED** [review]
Noticed a couple more small things in `setup.cfg` but looking pretty good.

Also, you should probably update the build command in `make package_woodwork` to remove references to `setup.py`. See FT for reference: https://github.com/alteryx/featuretools/blob/aaa84e3634988b2e6d2c528ff240a397655024d7/Makefile#L44

File: setup.cfg (hunk 11 above)
Comment (jeff-hernandez) [reply setup.cfg]: updated

File: setup.cfg (hunk 12 above)
Comment (jeff-hernandez) [reply setup.cfg:20]: updated

**Review by thehomebrewnerd — COMMENTED** [review]
Looking good, but I think we still should update the makefile package command to remove references to setup.py.

**Review by thehomebrewnerd — APPROVED** [review]
LGTM!

**Pull request merged**